python/seldon_core/seldon_client.py

```diff
@@ -0,0 +1,1394 @@
+    json_to_seldon_message, feedback_to_json, seldon_messages_to_json
+import numpy as np
+import grpc
+import requests
+from requests.auth import HTTPBasicAuth
+from typing import Tuple, Dict, Union, List, Optional
+import json
+import logging
+
+logger = logging.getLogger(__name__)
+
+class SeldonClientException(Exception):
+    """
+    Seldon Client Exception
+    """
+    status_code = 400
+
+    def __init__(self, message):
+        Exception.__init__(self)
+        self.message = message
+
+
+class SeldonClientPrediction(object):
+    """
+    Data class to return from Seldon Client
+    """
+
+    def __init__(self, request: Optional[prediction_pb2.SeldonMessage], response: Optional[prediction_pb2.SeldonMessage],
+                 success: bool = True, msg: str = ""):
+        self.request = request
+        self.response = response
+        self.success = success
+        self.msg = msg
+
+    def __repr__(self):
+        return "Success:%s message:%s\nRequest:\n%s\nResponse:\n%s" % (
+            self.success, self.msg, self.request, self.response)
+
+
+class SeldonClientFeedback(object):
+    """
+    Data class to return from Seldon Client for feedback calls
+    """
+    def __init__(self, request: Union[prediction_pb2.Feedback], response: Union[prediction_pb2.SeldonMessage],
+                 success: bool = True,
+                 msg: str = ""):
+        self.request = request
+        self.response = response
+        self.success = success
+        self.msg = msg
+
+    def __repr__(self):
+        return "Success:%s message:%s\nRequest:\n%s\nResponse:\n%s" % (
+            self.success, self.msg, self.request, self.response)
+
+
+class SeldonClientCombine(object):
+    """
+    Data class to return from Seldon Client for aggregate calls
+    """
+    def __init__(self, request: Optional[prediction_pb2.SeldonMessageList], response: Optional[prediction_pb2.SeldonMessage],
+                 success: bool = True, msg: str = ""):
+        self.request = request
+        self.response = response
+        self.success = success
+        self.msg = msg
+
+    def __repr__(self):
+        return "Success:%s message:%s\nRequest:\n%s\nResponse:\n%s" % (
+            self.success, self.msg, self.request, self.response)
+
+
+class SeldonClient(object):
+    """
+    A reference Seldon API Client
+    """
+
+    def __init__(self, gateway: str = "ambassador", transport: str = "rest", namespace: str = None, deployment_name: str = None,
+                 payload_type: str = "tensor", oauth_key: str = None, oauth_secret: str = None,
+                 seldon_rest_endpoint: str = "localhost:8002", seldon_grpc_endpoint: str = "localhost:8004",
+                 ambassador_endpoint: str = "localhost:8003", microservice_endpoint: str = "localhost:5000",
+                 grpc_max_send_message_length: int =4 * 1024 * 1024, grpc_max_receive_message_length: int =4 * 1024 * 1024):
```

Comment:
Not for this PR but maybe it's worth to refactor the client to inject in it 1) authentication method (instead of addressing each type of authentication with "ifs"), 2) endpoint type (REST or GRPC).

This might help to standardise functionality or different types, drastically reduce the number of "ifs" and number of parameters.

Reply:
Another option could be to create a parent SeldonClient and inherit 2 classes: SeldonClientGRPC and SeldonClientRest and put them in 3 separate modules. Currently, `seldon_client.py` is huge with lots of similar bits.

Reply:
Yes, agree. Maybe leave for a future PR.